File: src/test/java/com/koddy/server/member/application/usecase/GetMentorReservedScheduleUseCaseTest.java
package com.koddy.server.member.application.usecase;

import com.koddy.server.coffeechat.domain.model.CoffeeChat;
import com.koddy.server.coffeechat.domain.repository.query.MentorReservedScheduleQueryRepository;
import com.koddy.server.common.UnitTest;
import com.koddy.server.common.fixture.CoffeeChatFixture.MentorFlow;
import com.koddy.server.member.application.usecase.query.GetMentorReservedSchedule;
import com.koddy.server.member.application.usecase.query.response.MentorReservedSchedule;
import com.koddy.server.member.domain.model.mentee.Mentee;
import com.koddy.server.member.domain.model.mentor.Mentor;
import com.koddy.server.member.domain.model.response.ScheduleResponse;
import com.koddy.server.member.domain.repository.MentorRepository;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.time.LocalDateTime;
import java.util.List;

import static com.koddy.server.common.fixture.MenteeFixture.MENTEE_1;
import static com.koddy.server.common.fixture.MenteeFixture.MENTEE_2;
import static com.koddy.server.common.fixture.MenteeFixture.MENTEE_3;
import static com.koddy.server.common.fixture.MentorFixture.MENTOR_1;
import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertAll;
import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;

@DisplayName("Member -> GetMentorReservedScheduleUseCase 테스트")
class GetMentorReservedScheduleUseCaseTest extends UnitTest {
    private final MentorRepository mentorRepository = mock(MentorRepository.class);
    private final MentorReservedScheduleQueryRepository mentorReservedScheduleQueryRepository = mock(MentorReservedScheduleQueryRepository.class);
    private final GetMentorReservedScheduleUseCase sut = new GetMentorReservedScheduleUseCase(mentorRepository, mentorReservedScheduleQueryRepository);

    private final Mentor mentor = MENTOR_1.toDomain().apply(1L);
    private final Mentee menteeA = MENTEE_1.toDomain().apply(2L);
    private final Mentee menteeB = MENTEE_2.toDomain().apply(3L);
    private final Mentee menteeC = MENTEE_3.toDomain().apply(4L);

    @Test
    @DisplayName("특정 Year-Month에 대해서 멘토의 예약된 스케줄 정보를 조회한다")
    void success() {
        // given
        final GetMentorReservedSchedule query = new GetMentorReservedSchedule(mentor.getId(), 2024, 2);
        given(mentorRepository.getByIdWithSchedules(query.mentorId())).willReturn(mentor);

        final LocalDateTime startB = LocalDateTime.of(2024, 2, 18, 18, 0);
        final LocalDateTime startC = LocalDateTime.of(2024, 2, 15, 18, 0);

        final CoffeeChat coffeeChatA = MentorFlow.suggestAndReject(mentor, menteeA).apply(1L);
        final CoffeeChat coffeeChatB = MentorFlow.suggestAndPending(startB, startB.plusMinutes(30), mentor, menteeB).apply(2L);
        final CoffeeChat coffeeChatC = MentorFlow.suggestAndPending(startC, startC.plusMinutes(30), mentor, menteeC).apply(2L);
        given(mentorReservedScheduleQueryRepository.fetchReservedCoffeeChat(query.mentorId(), query.year(), query.month())).willReturn(List.of(coffeeChatC, coffeeChatB));

        // when
        final MentorReservedSchedule result = sut.invoke(query);

        // then
        assertAll(
                () -> verify(mentorRepository, times(1)).getByIdWithSchedules(query.mentorId()),
                () -> verify(mentorReservedScheduleQueryRepository, times(1)).fetchReservedCoffeeChat(query.mentorId(), query.year(), query.month()),
                () -> assertThat(result.period().startDate()).isEqualTo(mentor.getMentoringPeriod().getStartDate()),
                () -> assertThat(result.period().endDate()).isEqualTo(mentor.getMentoringPeriod().getEndDate()),
                () -> assertThat(result.schedules()).containsExactlyInAnyOrderElementsOf(
                        mentor.getSchedules()
                                .stream()
                                .map(it -> ScheduleResponse.from(it.getTimeline()))
                                .toList()
                ),
                () -> assertThat(result.timeUnit()).isEqualTo(mentor.getMentoringTimeUnit()),
                () -> assertThat(result.reserved())
                        .map(MentorReservedSchedule.Reserved::start)
                        .containsExactly(coffeeChatC.getReservation().getStart(), coffeeChatB.getReservation().getStart()),
                () -> assertThat(result.reserved())
                        .map(MentorReservedSchedule.Reserved::end)
                        .containsExactly(coffeeChatC.getReservation().getEnd(), coffeeChatB.getReservation().getEnd())
        );
    }
}

File: src/test/kotlin/com/koddy/server/member/application/usecase/GetMentorReservedScheduleUseCaseTest.kt
package com.koddy.server.member.application.usecase

import com.koddy.server.coffeechat.domain.model.CoffeeChat
import com.koddy.server.coffeechat.domain.repository.query.MentorReservedScheduleQueryRepository
import com.koddy.server.common.UnitTestKt
import com.koddy.server.common.fixture.MenteeFixtureStore.menteeFixture
import com.koddy.server.common.fixture.MentorFixtureStore.mentorFixture
import com.koddy.server.common.fixture.MentorFlow
import com.koddy.server.member.application.usecase.query.GetMentorReservedSchedule
import com.koddy.server.member.application.usecase.query.response.MentorReservedSchedule
import com.koddy.server.member.domain.model.mentee.Mentee
import com.koddy.server.member.domain.model.mentor.Mentor
import com.koddy.server.member.domain.service.MemberReader
import io.kotest.assertions.assertSoftly
import io.kotest.core.annotation.DisplayName
import io.kotest.core.spec.style.FeatureSpec
import io.kotest.matchers.collections.shouldContainExactly
import io.kotest.matchers.shouldBe
import io.mockk.every
import io.mockk.mockk
import java.time.LocalDateTime

@UnitTestKt
@DisplayName("Member -> GetMentorReservedScheduleUseCase 테스트")
internal class GetMentorReservedScheduleUseCaseTest : FeatureSpec({
    val memberReader = mockk<MemberReader>()
    val mentorReservedScheduleQueryRepository = mockk<MentorReservedScheduleQueryRepository>()
    val sut = GetMentorReservedScheduleUseCase(
        memberReader,
        mentorReservedScheduleQueryRepository,
    )

    val mentor: Mentor = mentorFixture(id = 1L).toDomain()
    val menteeA: Mentee = menteeFixture(id = 2L).toDomain()
    val menteeB: Mentee = menteeFixture(id = 3L).toDomain()
    val menteeC: Mentee = menteeFixture(id = 4L).toDomain()

    feature("GetMentorReservedScheduleUseCase's invoke") {
        scenario("특정 Year/Month에 대해서 멘토의 예약된 스케줄 정보를 조회한다") {
            val query = GetMentorReservedSchedule(mentor.id, 2024, 2)
            every { memberReader.getMentorWithSchedules(query.mentorId) } returns mentor

            val start1 = LocalDateTime.of(2024, 2, 18, 18, 0)
            val start2 = LocalDateTime.of(2024, 2, 15, 18, 0)
            val coffeeChats: List<CoffeeChat> = listOf(
                MentorFlow.suggestAndReject(id = 1L, mentor = mentor, mentee = menteeA),
                MentorFlow.suggestAndPending(id = 2L, start = start1, end = start1.plusMinutes(30), mentor = mentor, mentee = menteeB),
                MentorFlow.suggestAndPending(id = 3L, start = start2, end = start2.plusMinutes(30), mentor = mentor, mentee = menteeC),
            )
            every {
                mentorReservedScheduleQueryRepository.fetchReservedCoffeeChat(
                    query.mentorId,
                    query.year,
                    query.month,
                )
            } returns listOf(
                coffeeChats[2],
                coffeeChats[1],
            )

            val result: MentorReservedSchedule = sut.invoke(query)
            assertSoftly(result) {
                period!!.startDate shouldBe mentor.mentoringPeriod!!.startDate
                period!!.endDate shouldBe mentor.mentoringPeriod!!.endDate
                schedules.map { it.dayOfWeek } shouldContainExactly mentor.schedules.map { it.timeline.dayOfWeek.kor }
                schedules.map { it.start.hour } shouldContainExactly mentor.schedules.map { it.timeline.startTime.hour }
                schedules.map { it.start.minute } shouldContainExactly mentor.schedules.map { it.timeline.startTime.minute }
                schedules.map { it.end.hour } shouldContainExactly mentor.schedules.map { it.timeline.endTime.hour }
                schedules.map { it.end.minute } shouldContainExactly mentor.schedules.map { it.timeline.endTime.minute }
                timeUnit shouldBe mentor.mentoringTimeUnit
                reserved.map { it.start } shouldContainExactly listOf(
                    coffeeChats[2].reservation!!.start,
                    coffeeChats[1].reservation!!.start,
                )
                reserved.map { it.end } shouldContainExactly listOf(
                    coffeeChats[2].reservation!!.end,
                    coffeeChats[1].reservation!!.end,
                )
            }
        }
    }
})
